### PR TITLE
Prevent status flap when re-registering a check

### PR DESCRIPTION
**Pull request by Aestek**
Fixes point `#2` of: https://github.com/hashicorp/consul/issues/4903

When registering a service each healthcheck status is saved and restored (https://github.com/hashicorp/consul/blob/master/agent/agent.go#L1914) to avoid unnecessary flaps in health state.
This change extends this feature to single check registration by moving this protection in `AddCheck()` so that both `PUT /v1/agent/service/register` and `PUT /v1/agent/check/register` behave in the same idempotent way.

#### Steps to reproduce
1. Register a check :
```
curl -X PUT \
  http://127.0.0.1:8500/v1/agent/check/register \
  -H 'Content-Type: application/json' \
  -d '{
  "Name": "my_check",
  "ServiceID": "srv",
  "Interval": "10s",
  "Args": ["true"]
}'
```
2. The check will initialize and change to `passing`
3. Run the same request again
4. The check status will quickly go from `critical` to `passing` (the delay for this transission is determined by https://github.com/hashicorp/consul/blob/master/agent/checks/check.go#L95)

### CODE REVIEW

**Comment by pierresouchay** [comment]
@banks while https://github.com/hashicorp/consul/issues/4905 is harder to merge, this one is simple enough and just improve the situation by not resetting the state and avoid unnecessary flap of services, would be a nice improvement IMHO

File: agent/agent.go [hunk 1]
```diff
@@ -2052,6 +2047,14 @@ func (a *Agent) AddCheck(check *structs.HealthCheck, chkType *structs.CheckType,
...
 	defer a.checkLock.Unlock()
 
+	// snapshot the current state of the health check to avoid potential flapping
+	existing := a.State.Check(check.CheckID)
+	defer func() {
+		if existing != nil {
+			a.State.UpdateCheck(check.CheckID, existing.Status, existing.Output)
+		}
+	}()
+
```
Comment (banks) [inline agent/agent.go]: It's not obvious to me why this is different from the old code that takes a snapshot before any of this and then restores all checks in that snapshot afterwards?

**Comment by banks** [comment]
Thanks @Aestek @pierresouchay, I've added it to the list to consider first glance seems good.

To make sure I understand, The code that is removed here already means that `PUT /v1/agent/service/register` behaves correctly right? It's only the `PUT /v1/agent/check/register` that exhibits the flap bug?

**Comment by Aestek** [comment]
@banks yes, `/v1/agent/service/register` was already protected against flaps, this change extends the protection to `/v1/agent/check/register`

**Comment by pierresouchay** [comment]
@mkeeler thank you

**Pull request merged**